### PR TITLE
[VarDumper] Add interval caster

### PR DESCRIPTION
**Pull request by maidmaid**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | /
| License       | MIT
| Doc PR        | /

 cf https://github.com/symfony/symfony/pull/22431#issuecomment-312567747

### CODE REVIEW

**Comment by maidmaid** [comment]
First proposal:

- the main public properties (``y``, ``m``, ``d``, ``h``, ``i``, ``s``, and ``f`` on PHP 7.1) are merged in oneline ``interval`` virtual property;
- the others public properties (``invert`` and ``days``) are unmodified;
- and the public properties not officially documentated (``weekday``, ``weekday_behavior``, ``first_last_day_of``, ``special_type``, ``special_amount``, ``have_weekday_relative``, ``have_special_relative``) are removed (see http://php.net/manual/en/class.dateinterval.php#dateinterval.synopsis).

| Before        | After           |
| ------------- | ------------- |
|  |  |

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 1]
```diff
@@ -38,4 +38,36 @@ public static function castDateTime(\DateTimeInterface $d, array $a, Stub $stub,
 
         return $a;
     }
+
+    public static function castInterval(\DateInterval $i, array $a, Stub $stub, $isNested, $filter)
+    {
+        $prefix = Caster::PREFIX_VIRTUAL;
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: no need to define a variable, as it is used only once

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 2]
```diff
@@ -38,4 +38,36 @@ public static function castDateTime(\DateTimeInterface $d, array $a, Stub $stub,
...
+
+    private static function formatInterval(\DateInterval $i)
+    {
+        $format = '%R '
+            .($i->y ? '%yy ' : '')
+            .($i->m ? '%mm ' : '')
+            .($i->d ? '%dd ' : '')
+        ;
+
+        if (PHP_VERSION_ID >= 70100) {
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: please use `\PHP_VERSION_ID`, to let the version comparison be applied at compile time (not qualifying the constant requires doing it at runtime to check the constant first in the current namespace)

File: src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php [hunk 3]
```diff
@@ -84,4 +84,48 @@ public function provideDateTimes()
...
+  +"days": false
+}
+EODUMP;
+
+        $this->assertDumpMatchesFormat($xDump, $interval);
+    }
+
+    public function provideIntervals()
+    {
+        $ms = function () { return PHP_VERSION_ID >= 70100 ? '.000000' : ''; };
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php]: no need to use a function. Use a simple variable, as this is purely static.

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 1 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Fixed

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 2 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: ``\`` added

File: src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php (hunk 3 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php]: Updated

**Comment by ogizanagi** [comment]
I agree with @nicolas-grekas's original comment (https://github.com/symfony/symfony/pull/22431#issuecomment-312567747), so I'm not convinced by the new suggested output. 
Perhaps we could just exclude non-officially documented public properties and add your `interval` virtual property. Then, potentially exclude public properties merged in the `interval` virtual properties when used with the `EXCLUDE_VERBOSE` filter, so it won't be shown within the form and validator data collectors for instance, but still shown when dumping around elsewhere.

Also, why showing the `invert` field when it's already part of the `interval`? According to the above suggestions, I'll hide it too when using `EXCLUDE_VERBOSE`.

Would it be interesting to also output it as the [time_interval spec](http://php.net/manual/en/dateinterval.construct.php)?
https://github.com/jaroslavtyc/granam-date-interval/blob/master/Granam/DateInterval/DateInterval.php#L175-L200

WDYT?

**Comment by maidmaid** [comment]
Thank you for your thoughts @ogizanagi. I would say this:

- Why do not you like removing time properties? This deletion is in favor of ``interval`` that is a full representation of this props.
- About ``invert`` prop, you are right. 
- About ``time_interval`` spec, I think that ``interval`` virtual prop is a good replacement and more intuitive to read (``P1Y2M3DT4H5M6S`` vs ``+ 1y 2m 3d 04:05:06``). Moreover, it does not handle milliseconds and the negative intervals.

**Comment by ogizanagi** [comment]
> Why do not you like removing time properties? This deletion is in favor of interval that is a full representation of this props.

Just quoting @nicolas-grekas:

> Dump should be informational about how things actually works or can be used. On a DateInterval, the listed values are actual public properties. VarDumper is about showing actual facts about objects, not alternate ones

I'm using [psysh](http://psysh.org/) quite often, to quickly test or demonstrate things, and despite built-in introspection/doc commands like `ls -al`, `show`, ... I like the usage of the var-dumper component to output everything and the fact it shows real internals, not representations. I wish it stay this way, but I'm not against adding some helpful representation in addition.

**Comment by nicolas-grekas** [comment]
Same opinion as @ogizanagi.
Maybe display the short output when the EXCLUDE_VERBOSE flag is set (the default on psysh)?

**Comment by maidmaid** [comment]
Ok, here the new result according to your comments:
 
| normal        | ``EXCLUDE_VERBOSE``           |
| ------------- | ------------- |
|  |  |

**Comment by nicolas-grekas** [comment]
"normal" should display all existing properties, even undocumented ones. Also, why keep "days" in compact mode?

**Comment by maidmaid** [comment]
It's updated: "normal" displays all, and only ``interval`` prop is keeped in compact mode.

File: src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php [hunk 4]
```diff
@@ -84,4 +85,64 @@ public function provideDateTimes()
             array('2017-04-30 00:00:00.000000', '+02:00', '2017-04-30 00:00:00.000000 +02:00'),
         );
     }
+
+    /**
+     * @dataProvider provideIntervals
+     */
+    public function testCastInterval($interval_spec, $invert, $expected)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php]: should be `$intervalSpec`

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 5]
```diff
@@ -38,4 +38,30 @@ public static function castDateTime(\DateTimeInterface $d, array $a, Stub $stub,
 
         return $a;
     }
+
+    public static function castInterval(\DateInterval $interval, array $a, Stub $stub, $isNested, $filter)
+    {
+        $i = array(Caster::PREFIX_VIRTUAL.'interval' => self::formatInterval($interval));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: what about using a ConstStub here that gives the number of seconds on hover (html mode):
`new ConstStub(self::formatInterval($interval), $numberOfSeconds);`

**Review by nicolas-grekas — APPROVED** [review]
with some suggestions

File: src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php (hunk 4 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php]: Fixed (in [official doc](http://php.net/manual/fr/dateinterval.construct.php), this argument is named ``$interval_spec``).

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 5 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Good idea. Number of seconds added.

**Comment by maidmaid** [comment]
How to fix kind of HHVM failure?
```
1) Symfony\Component\VarDumper\Tests\Caster\DateCasterTest::testDumpInterval with data set #0 ('PT0S', 0, '0s', '0s')
Failed asserting that format description matches text.
--- Expected
+++ Actual
@@ @@
 DateInterval {
   interval: 0s
 }
-}
```

**Comment by nicolas-grekas** [comment]
```diff
diff --git a/src/Symfony/Component/VarDumper/Caster/DateCaster.php b/src/Symfony/Component/VarDumper/Caster/DateCaster.php
index c9edc10..1ecd962 100644
--- a/src/Symfony/Component/VarDumper/Caster/DateCaster.php
+++ b/src/Symfony/Component/VarDumper/Caster/DateCaster.php
@@ -58,7 +58,7 @@ class DateCaster
             .($i->d ? '%dd ' : '')
         ;
 
-        if (\PHP_VERSION_ID >= 70100) {
+        if (isset($i->f)) {
             $format .= $i->h || $i->i || $i->s || $i->f ? '%H:%I:%S.%F' : '';
         } else {
             $format .= $i->h || $i->i || $i->s ? '%H:%I:%S' : '';
diff --git a/src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php b/src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php
index 9db333e..c4c1263 100644
--- a/src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php
+++ b/src/Symfony/Component/VarDumper/Tests/Caster/DateCasterTest.php
@@ -97,8 +97,7 @@ EODUMP;
         $xDump = <<<EODUMP
 DateInterval {
   interval: $expected
-%a
-}
+%A}
 EODUMP;
 
         $this->assertDumpMatchesFormat($xDump, $interval);
@@ -162,7 +161,8 @@ EODUMP;
 
     public function provideIntervals()
     {
-        $ms = \PHP_VERSION_ID >= 70100 ? '.000000' : '';
+        $interval = new \DateInterval('PT0S');
+        $ms = isset($interval->f);
 
         return array(
             array('PT0S', 0, '0s', '0s'),
```

**Comment by maidmaid** [comment]
Mmh, ``isset($interval->f)`` triggers a notice in 7.0, see https://3v4l.org/hJB5A. I use ``PHP_VERSION_ID``?

**Comment by nicolas-grekas** [comment]
but my HHVM fails when PHP_VERSION_ID is used...

**Comment by nicolas-grekas** [comment]
you should report the notice as a bug I guess

**Comment by nicolas-grekas** [comment]
using both PHP_VERSION_ID and the isset check should do it

**Comment by maidmaid** [comment]
There was no error related with ``PHP_VERSION_ID`` [previously](https://github.com/symfony/symfony/pull/23357#issuecomment-314719651). Tests seems fine now (failure seems unrelated).

**Comment by maidmaid** [comment]
> you should report the notice as a bug

https://bugs.php.net/bug.php?id=74911 ;)

**Comment by nicolas-grekas** [comment]
> was no error related with PHP_VERSION_ID previously

that's because the behavior of hhvm changes fro mversion to version and Travis doesn't have the latest. But trust me, this check is not enough, you need the isset one also :)

**Comment by nicolas-grekas** [comment]
(or install hhvm to see by yourself ;) )

**Comment by maidmaid** [comment]
Updated: used ``PHP_VERSION_ID`` and the isset check.

**Comment by maidmaid** [comment]
Time to merge? :)

**Comment by nicolas-grekas** [comment]
Thank you @maidmaid.

**Pull request merged**